File: src/session1/td/NatDecimal.java
package session1.td;

public class NatDecimal implements Nat{
	
	//attributs
	private String chiffres;
	public static final FabriqueNat<Nat> FAB = new NatDecimal("");
	
	//constructeur
	public NatDecimal(String rep) {
		this.chiffres=rep;
	}
	
	public Nat somme(Nat x) {
		 int t = this.taille() < x.taille() ? x.taille() : this.taille();
		  StringBuilder rep = new StringBuilder();
		  int retenue = 0;
		  for(int i = 0; i < t; i++){
		    int chiffre = this.chiffre(i) + x.chiffre(i) + retenue;
		    if(chiffre > 9){
		      chiffre = chiffre - 10;
		      retenue = 1;
		    }else{
		      retenue = 0;
		    }
		    rep.append(Integer.toString(chiffre));
		  }
		  rep = retenue == 0 ? rep : rep.append(1);
		  return new NatDecimal(rep.reverse().toString());
		}

	@Override
	public Nat creerNatAvecValeur(int x) {
		return new NatDecimal(String.valueOf(x));
	}

	@Override
	public NatDecimal creerZero() {
		return new NatDecimal("0");
	}

	@Override
	public boolean estNul() {
		return this.val()==0;
	}

	@Override
	public Nat predecesseur() {
		int t = this.taille();
		  StringBuilder rep = new StringBuilder();
		  int retenue = 0;
		  for(int i = t-1; i >= 0; i--){
		    int chiffre = this.chiffre(i) - 1 + retenue;
		    if(chiffre < 0){
		    	chiffre = 9;
		    	retenue = 0;
		    }else{
		      retenue = 1;
		    }
		    rep.append(Integer.toString(chiffre));
		  }
		  String nb = rep.toString();
		  //System.out.println(nb.charAt(nb.length()-1));
		  if(nb.charAt(nb.length()-1)=='0'&&nb.length()>1) {
			  nb = nb.substring(0, nb.length() - 1);
		  }
		  return new NatDecimal(new StringBuilder(nb).reverse().toString());
	}

	@Override
	public int val() {
		return Integer.parseInt(this.chiffres);
	}

	@Override
	public int taille() {
		return this.chiffres.length();
	}

	@Override
	public int chiffre(int i) {
		try {
			return Integer.parseInt(String.valueOf(this.chiffres.charAt(i)));
		} catch (Exception e) {
			return 0;		
		}
	}

	@Override
	public Nat creerSuccesseur(Nat predecesseur) {
		int t = predecesseur.taille();
		  StringBuilder rep = new StringBuilder();
		  int retenue = 1;
		  for(int i = 0; i < t; i++){
		    int chiffre = predecesseur.chiffre(i) + retenue;
		    if(chiffre > 9){
		      chiffre = chiffre - 10;
		      retenue = 1;
		    }else{
		      retenue = 0;
		    }
		    rep.append(Integer.toString(chiffre));
		  }
		  rep = retenue == 0 ? rep : rep.append(1);
		  return new NatDecimal(rep.reverse().toString());
	}

	@Override
	public Nat zero() {
		return this.creerZero();
	}

	@Override
	public Nat produit(Nat n) {
		if(this.estNul()) {
			return this.creerZero();
		}else {
			return this.creerNatAvecValeur(n.val() + n.produit(this.predecesseur()).val());
		}
	}

	@Override
	public Nat un() {
		return this.creerNatAvecValeur(1);
	}

	@Override
	public Nat modulo(Nat x) {
		if(this.val()<x.val()) {
			return this;
		}else {
			return this.creerNatAvecValeur(this.creerNatAvecValeur(this.val()-x.val()).modulo(x).val());
		}
		//return this.creerNatAvecValeur(this.val() % x.val());
	}

	@Override
	public Nat div(Nat x) {
		if(this.estNul()) {
			return this.creerZero();
		}else if (this.modulo(x).val()==this.creerZero().val()){
			return this.creerNatAvecValeur(1 + this.predecesseur().div(x).val());
		}else {
			return this.creerNatAvecValeur(this.predecesseur().div(x).val());
		}
	}

	@Override
	public String toString(){
		return this.chiffres;
	}
	
	@Override
	public boolean equals(Object o) {
		Nat x = (Nat)o;
		int t = this.taille() < x.taille() ? x.taille() : this.taille();
		boolean equal = true;
		  for(int i = 0; i < t; i++){
		    if(this.chiffre(i) != x.chiffre(i)){
		    	equal = false;
		    }
		  }
		  return equal;
	}

}

File: TD2/Typescript/session2/td/heritageMultiple/natDecimal.ts
import {
    Nat, FabriqueNat, AlgebreNatDecimal,AlgebreNatParInt, AlgebreNatRecursif
} from "./naturels";
import { assembleurString } from "../../../bibliotheque/assembleurChaine";

import {
    NombreDecimal
} from "./natInductif";

function nettoyage(s: string): string {
    const a = assembleurString(s);
    for (let i = 0; i < s.length; i++) {
        if (s.charAt(i) === "0") {
            a.retirerCaractereDebut();
        } else {
            break;
        }
    }
    return a.chaine();
}

export class NatDecimal extends NombreDecimal implements AlgebreNatDecimal {

    /*
    * Aucun contrôle n'est réalisé. L'argument est supposé correct :
    * - représentation décimale (suite de 0 et 1, pas de 0 en tête inutile).
    * Pour un contrôle, utiliser la fabrique.
    */
    constructor(chiffres: string) {super(chiffres); }

    creerNatAvecRepresentation(repDecimale: string): Nat {
        if (repDecimale.search(/^d/) !== -1) {
            throw new Error("* Erreur : représentation non décimale.");
        }
        repDecimale = nettoyage(repDecimale);
        if (repDecimale === "") {
            repDecimale = "0";
        }
        return new NatDecimal(repDecimale);


    }

    creerNatAvecValeur(valeur: number): Nat {
        return new NatDecimal(valeur.toString());
    }

    creerZero(): Nat {
        return new NatDecimal("0");
    }

    creerSuccesseur(predecesseur: Nat): Nat {
        let t = predecesseur.taille();
        const a = assembleurString("");
        let retenue = 1;
        for (let i = 0; i < t; i++) {
            let chiffre = predecesseur.chiffre(i) + retenue;
            if (chiffre > 9) {
                chiffre = chiffre - 10;
                retenue = 1;
            } else {
                retenue = 0;
            }
            a.ajouterCaractereDebut(chiffre.toString());
        }
        if (retenue === 1) {
            a.ajouterCaractereDebut("1");
        }
        return new NatDecimal(a.chaine());
    }


    estNul(): boolean {
        return this.val() == 0;
    }

    predecesseur(): Nat {
        if (this.estNul()) {
            throw new Error("* Erreur : naturel nul sans prédécesseur.");
        }
        const t = this.taille();
        const rep: string[] = [];
        let retenue = -1;
        for (let i = 0; i < t; i++) {
            let chiffre = this.chiffre(i) + retenue;
            if (chiffre === -1) {
                chiffre = 9;
                retenue = -1;
            } else {
                retenue = 0;
            }
            rep.push(chiffre.toString());
        }
        return this.creerNatAvecRepresentation(rep.reverse().join(""));
    }

    somme(x: Nat): Nat {
        const t = this.taille() < x.taille() ? x.taille() : this.taille();
        const rep: string[] = [];
        let retenue = 0;
        for (let i = 0; i < t; i++) {
            let chiffre = this.chiffre(i) + x.chiffre(i) + retenue;
            if (chiffre > 9) {
                chiffre = chiffre - 10;
                retenue = 1;
            } else {
                retenue = 0;
            }
            rep.push(chiffre.toString());
        }
        if (retenue === 1) {
            rep.push("1");
        }
        return this.creerNatAvecRepresentation(rep.reverse().join(""));
    }

    zero(): Nat {
        return this.creerZero();
    }

    produit(x: Nat): Nat {
        if (x.estEgal(DIX)) {
            return this.creerNatAvecRepresentation(this.representation() + "0");
        }
        let res = this.zero();
        let index = this.zero();
        while (!index.estEgal(x)) {
            res = res.somme(this);
            index = this.creerSuccesseur(index);
        }
        return res;
    }

    un(): Nat {
        return this.creerNatAvecRepresentation("1");
    }

    modulo(x: Nat): Nat {
        if (x.estEgal(DIX)) {
            return this.creerNatAvecValeur(this.chiffre(0));
        }
        let courant = this.zero();
        let r = this.zero();
        while (!courant.estEgal(this)) {
            r = this.creerSuccesseur(r);
            if (r.estEgal(x)) {
                r = this.zero();
            }
            courant = this.creerSuccesseur(courant);
        }
        return r;
    }

    div(x: Nat): Nat {
        if (x.estEgal(DIX)) {
            if (this.taille() == 1)
                return this.zero();
            return this.creerNatAvecRepresentation(
                this.representation().substring(0, this.taille() - 1));
        }
        let courant = this.zero();
        let q = this.zero();
        let r = this.zero();
        while (!courant.estEgal(this)) {
            r = this.creerSuccesseur(r);
            if (r.estEgal(x)) {
                r = this.zero();
                q = this.creerSuccesseur(q);
            }
            courant = this.creerSuccesseur(courant);
        }
        return q;
    }

    representation(): string {
        return super.representation();
    }

    estEgal(n: Nat): boolean {
        return this.representation() === n.representation();
    }


}

export const natDecimal: FabriqueNat<Nat> = new NatDecimal("0");

const DIX: Nat = natDecimal.creerNatAvecRepresentation("10");

export class NatDecimalParInt extends NombreDecimal implements AlgebreNatParInt {
  constructor(chiffres: string) {super(chiffres); }

  creerNatAvecRepresentation(repDecimale: string): Nat {
      if (repDecimale.search(/^d/) !== -1) {
          throw new Error("* Erreur : représentation non décimale.");
      }
      repDecimale = nettoyage(repDecimale);
      if (repDecimale === "") {
          repDecimale = "0";
      }
      return new NatDecimal(repDecimale);
  }

  toString() : string{
    return super.representation();
  }

}

export class NatDecimalRecursif extends NombreDecimal implements AlgebreNatRecursif {
  constructor(chiffres: string) {super(chiffres); }

  creerNatAvecRepresentation(repDecimale: string): Nat {
      if (repDecimale.search(/^d/) !== -1) {
          throw new Error("* Erreur : représentation non décimale.");
      }
      repDecimale = nettoyage(repDecimale);
      if (repDecimale === "") {
          repDecimale = "0";
      }
      return new NatDecimal(repDecimale);
  }

  toString() : string{
    return super.representation();
  }

}
